source/uwp/Renderer/lib/XamlBuilder.cpp

```diff
@@ -1123,6 +1123,243 @@ namespace AdaptiveCards { namespace Rendering { namespace Uwp
+        RETURN_IF_FAILED(run.As(&runAsInline));
+
+        RETURN_IF_FAILED(inlines->Append(runAsInline.Get()));
+        return S_OK;
+    }
+
+    HRESULT SetTextOnXamlTextBlock(
+        IAdaptiveRenderContext* renderContext,
+        HSTRING textIn,
+        ITextBlock * textBlock)
+    {
+        ComPtr<ABI::Windows::Data::Xml::Dom::IXmlDocument> xmlDocument = XamlHelpers::CreateXamlClass<ABI::Windows::Data::Xml::Dom::IXmlDocument>(HStringReference(RuntimeClass_Windows_Data_Xml_Dom_XmlDocument));
+
+        ComPtr<ABI::Windows::Data::Xml::Dom::IXmlDocumentIO> xmlDocumentIO;
+        RETURN_IF_FAILED(xmlDocument.As(&xmlDocumentIO));
+
+        HRESULT hr = xmlDocumentIO->LoadXml(textIn);
+        if (FAILED(hr))
+        {
+            RETURN_IF_FAILED(textBlock->put_Text(textIn));
+        }
+        else
+        {
+            ComPtr<IVector<ABI::Windows::UI::Xaml::Documents::Inline*>> inlines;
+            RETURN_IF_FAILED(textBlock->get_Inlines(inlines.GetAddressOf()));
+
+            ComPtr<ABI::Windows::Data::Xml::Dom::IXmlNode> xmlDocumentAsNode;
+            RETURN_IF_FAILED(xmlDocument.As(&xmlDocumentAsNode));
+
+            ComPtr<ABI::Windows::Data::Xml::Dom::IXmlNode> pNode;
+            RETURN_IF_FAILED(xmlDocumentAsNode->get_FirstChild(&pNode));
+
+            ComPtr<ABI::Windows::Data::Xml::Dom::IXmlNode> childNode;
+            RETURN_IF_FAILED(pNode->get_FirstChild(&childNode));
+
+            while (childNode != nullptr)
+            {
+                HString nodeName;
+                RETURN_IF_FAILED(childNode->get_NodeName(nodeName.GetAddressOf()));
+
+                INT32 isOrderedListResult;
+                RETURN_IF_FAILED(WindowsCompareStringOrdinal(nodeName.Get(), HStringReference(L"ol").Get(), &isOrderedListResult));
+
+                INT32 isUnorderedListResult;
+                RETURN_IF_FAILED(WindowsCompareStringOrdinal(nodeName.Get(), HStringReference(L"ul").Get(), &isUnorderedListResult));
+
+                INT32 isLinkResult;
+                RETURN_IF_FAILED(WindowsCompareStringOrdinal(nodeName.Get(), HStringReference(L"a").Get(), &isLinkResult));
+
+                INT32 isBoldResult;
+                RETURN_IF_FAILED(WindowsCompareStringOrdinal(nodeName.Get(), HStringReference(L"strong").Get(), &isBoldResult));
+
+                INT32 isItalicResult;
+                RETURN_IF_FAILED(WindowsCompareStringOrdinal(nodeName.Get(), HStringReference(L"em").Get(), &isItalicResult));
+
+                bool nodeHandled = false;
+                if (isOrderedListResult == 0 || isUnorderedListResult == 0)
+                {
+                    nodeHandled = SUCCEEDED(AddListInlines(childNode.Get(), (isOrderedListResult == 0), inlines.Get()));
+                }
+                else if (isLinkResult == 0)
+                {
+                    nodeHandled = SUCCEEDED(AddLinkInline(childNode.Get(), inlines.Get()));
+                }
+                
+                if (!nodeHandled)
```

Comment:
>if (!nodeHandled) [](start = 16, length = 17)

What do you think of using an else if for this one as well? we can keep the !nodeHandled as an error case when one of the AddXXInline functions report failure. In this version if AddList or AddLink fail, it will still try to AddText

Reply:
Yeah, that was my intent.  The scenario i ran into was that the link was missing it's href (i had manually added the HTML incorrectly).  At the time i was thinking that it would be better to display the malformed html as text than fail to render the card.  That said, now that i think about it the html is something we're generating (not a card author's mistake) so maybe it's fine to fail out here if it's bogus for some reason.

---
In reply to: [153663351](https://github.com/Microsoft/AdaptiveCards/pull/990#discussion_r153663351) [](ancestors = 153663351)